File: avro/avro-annotation-processor/src/main/java/io/koraframework/avro/annotation/processor/writer/AvroWriterGenerator.java
package io.koraframework.avro.annotation.processor.writer;

import com.palantir.javapoet.*;
import io.koraframework.annotation.processor.common.CommonClassNames;
import io.koraframework.annotation.processor.common.CommonUtils;
import io.koraframework.avro.annotation.processor.AvroTypes;
import io.koraframework.avro.annotation.processor.AvroUtils;
import org.jspecify.annotations.Nullable;

import javax.annotation.processing.ProcessingEnvironment;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import java.io.ByteArrayOutputStream;
import java.io.IOException;

public class AvroWriterGenerator {
    private final ProcessingEnvironment env;

    public AvroWriterGenerator(ProcessingEnvironment processingEnvironment) {
        this.env = processingEnvironment;
    }

    public void generate(TypeElement element) {
        var typeName = TypeName.get(element.asType());
        var typeBuilder = TypeSpec.classBuilder(AvroUtils.writerName(element))
            .addAnnotation(AnnotationSpec.builder(CommonClassNames.koraGenerated)
                .addMember("value", CodeBlock.of("$S", AvroWriterGenerator.class.getCanonicalName()))
                .build())
            .addAnnotation(AvroTypes.avro)
            .addSuperinterface(ParameterizedTypeName.get(AvroTypes.writer, typeName))
            .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
            .addOriginatingElement(element);

        for (var typeParameter : element.getTypeParameters()) {
            typeBuilder.addTypeVariable(TypeVariableName.get(typeParameter));
        }

        typeBuilder.addField(FieldSpec.builder(ArrayTypeName.of(TypeName.BYTE), "EMPTY")
            .addModifiers(Modifier.PRIVATE, Modifier.STATIC, Modifier.FINAL)
            .initializer("new byte[]{}")
            .build());
        typeBuilder.addField(FieldSpec.builder(AvroTypes.schema, "SCHEMA")
            .addModifiers(Modifier.PRIVATE, Modifier.STATIC, Modifier.FINAL)
            .initializer("$T.getClassSchema()", typeName)
            .build());
        typeBuilder.addField(FieldSpec.builder(AvroTypes.specificData, "SPECIFIC_DATA")
            .addModifiers(Modifier.PRIVATE, Modifier.STATIC, Modifier.FINAL)
            .initializer("new $T().getSpecificData()", typeName)
            .build());
        typeBuilder.addField(FieldSpec.builder(ParameterizedTypeName.get(AvroTypes.datumWriter, typeName), "WRITER")
            .addModifiers(Modifier.PRIVATE, Modifier.STATIC, Modifier.FINAL)
            .initializer("new $T<>(SCHEMA, SPECIFIC_DATA)", AvroTypes.datumWriter)
            .build());

        var method = MethodSpec.methodBuilder("writeBytes")
            .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
            .addException(IOException.class)
            .addAnnotation(Override.class)
            .addParameter(ParameterSpec.builder(typeName, "value").addAnnotation(Nullable.class).build())
            .returns(ArrayTypeName.of(TypeName.BYTE));
        method.beginControlFlow("if (value == null)");
        method.addStatement("return EMPTY");
        method.endControlFlow();
        method.beginControlFlow("try (var os = new $T())", ByteArrayOutputStream.class);
        method.addStatement("var encoder = $T.get().directBinaryEncoder(os, null)", AvroTypes.encoderFactory);
        method.addStatement("WRITER.write(value, encoder)");
        method.addStatement("encoder.flush()");
        method.addStatement("return os.toByteArray()");
        method.endControlFlow();

        typeBuilder.addMethod(method.build());
        var javaFile = JavaFile.builder(AvroUtils.classPackage(this.env.getElementUtils(), element), typeBuilder.build()).build();
        CommonUtils.safeWriteTo(this.env, javaFile);
    }
}

File: avro/avro-symbol-processor/src/main/kotlin/io/koraframework/avro/symbol/processor/writer/AvroWriterGenerator.kt
package io.koraframework.avro.symbol.processor.writer

import com.google.devtools.ksp.processing.CodeGenerator
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.jvm.throws
import com.squareup.kotlinpoet.ksp.addOriginatingKSFile
import com.squareup.kotlinpoet.ksp.toTypeParameterResolver
import com.squareup.kotlinpoet.ksp.toTypeVariableName
import com.squareup.kotlinpoet.ksp.writeTo
import io.koraframework.avro.symbol.processor.AvroTypes
import io.koraframework.avro.symbol.processor.classPackage
import io.koraframework.avro.symbol.processor.writerName
import io.koraframework.ksp.common.KspCommonUtils.generated
import io.koraframework.ksp.common.KspCommonUtils.toTypeName
import java.io.ByteArrayOutputStream
import java.io.IOException

class AvroWriterGenerator(val resolver: Resolver, val codeGenerator: CodeGenerator) {
    fun generate(declaration: KSClassDeclaration) {
        val typeName = declaration.toTypeName()
        val typeParameterResolver = declaration.typeParameters.toTypeParameterResolver()
        val writerInterface = AvroTypes.writer.parameterizedBy(typeName)
        val typeBuilder = TypeSpec.classBuilder(declaration.writerName())
            .generated(AvroWriterGenerator::class)
            .addAnnotation(AvroTypes.avro)
        declaration.containingFile?.let { typeBuilder.addOriginatingKSFile(it) }
        typeBuilder.addSuperinterface(writerInterface)
        declaration.typeParameters.forEach { typeBuilder.addTypeVariable(it.toTypeVariableName(typeParameterResolver)) }

        typeBuilder.addProperty(PropertySpec.builder("EMPTY", ByteArray::class).addModifiers(KModifier.PRIVATE, KModifier.FINAL).initializer("%T(0)", ByteArray::class).build())
        typeBuilder.addProperty(PropertySpec.builder("SCHEMA", AvroTypes.schema).addModifiers(KModifier.PRIVATE, KModifier.FINAL).initializer("%T.getClassSchema()", typeName).build())
        typeBuilder.addProperty(PropertySpec.builder("SPECIFIC_DATA", AvroTypes.specificData).addModifiers(KModifier.PRIVATE, KModifier.FINAL).initializer("%T().getSpecificData()", typeName).build())
        typeBuilder.addProperty(PropertySpec.builder("WRITER", AvroTypes.datumWriter.parameterizedBy(typeName)).addModifiers(KModifier.PRIVATE, KModifier.FINAL).initializer("%T(SCHEMA, SPECIFIC_DATA)", AvroTypes.datumWriter).build())

        val method = FunSpec.builder("writeBytes")
            .addModifiers(KModifier.PUBLIC, KModifier.FINAL, KModifier.OVERRIDE)
            .throws(IOException::class)
            .addParameter("value", typeName.copy(true))
            .returns(ByteArray::class)
        method.beginControlFlow("if (value == null)")
        method.addStatement("return EMPTY")
        method.endControlFlow()
        method.beginControlFlow("return %T().%M", ByteArrayOutputStream::class, MemberName("kotlin.io", "use"))
        method.addStatement("val encoder = %T.get().directBinaryEncoder(it, null)", AvroTypes.encoderFactory)
        method.addStatement("WRITER.write(value, encoder)", typeName)
        method.addStatement("encoder.flush()")
        method.addStatement("it.toByteArray()")
        method.endControlFlow()
        typeBuilder.addFunction(method.build())

        val spec = typeBuilder.build()
        FileSpec.builder(classPackage(declaration), spec.name!!).addType(spec).build().writeTo(codeGenerator = codeGenerator, aggregating = false)
    }
}
